extension/src/completion/autocompletionProvider.ts

```diff
@@ -171,30 +187,222 @@ function getMatchingWord(document: vscode.TextDocument, position: vscode.Positio
+                    this.previousItems.unshift(ucAtom); // most likely a '(' but defuns aren't required to have a (arg / local) declaration space so it could be nil
+                } else if (this.previousItems[0] && this.previousItems[0] === "SETQ") {
+                    context = 2; // this context tells me I am looking for a variable declaration
+                    this.tryAddSymbol(atom.symbol, AutoLispKind.Global);
+                } else {
+                    this.previousItems.unshift(ucAtom);
+                }
+            }
+        });
+    }
+
+    load(document: vscode.TextDocument, position: vscode.Position): void {
+        // get the scope of this position  
+        let kwRegex: RegExp = /\(\s*(DEFUN|DEFUN-Q)\s+\S{1,99}(\s|\(|;)/gi;
+        let tmp = findContainers(document, position);
+        let pr1: ElementRange = tmp.containerBlockComment;
+        let pr2: ElementRange[] = tmp.containerParens;
+        let contents: string = "";
+        const containers: ElementRange[] = findContainers(document, position).containerParens;
+        for (let ci = 0; ci < containers.length; ci++) {
+            const entry: ElementRange = containers[ci];
+            const value: string = document.getText(new vscode.Range(document.positionAt(entry.startPos.offsetInDocument), document.positionAt(entry.endPos.offsetInDocument)));
+            // retrieve the string contents of the first Defun scope
+            if (value.search(kwRegex) !== -1){                
+                contents = value;
+                break;
+            }
+        }
+        // Fallback for when the typed values are not in a Defun
+        if (contents === ""){
+            contents = document.getText(new vscode.Range(document.positionAt(containers.slice(-1)[0].startPos.offsetInDocument), document.positionAt(containers.slice(-1)[0].endPos.offsetInDocument)));
+        }
+        
+        
+        // Create and load that scope into a temporary document for contextual parsing, had to add this openFromContent method to narrow the LispParser SOW
+        const tempDoc: ReadonlyDocument = ReadonlyDocument.openFromContent(contents, "autolisp");
+        const lp: LispParser = new LispParser(tempDoc);        
+        lp.tokenizeString(contents, 0); 
+        // this atoms forest should only be comprised of a single Sexpression unless
+        if (lp.atomsForest.length === 1 && lp.atomsForest[0] instanceof Sexpression){
+            this.traverseSexpression(lp.atomsForest[0] as Sexpression);
+        } else {
+            // some kind of structural document error came into play
+            debugger;
+        }
+        this.previousItems.length = 0; // cleanup the tracking mess
+
+        // quick scan any open documents for defun headers. 
+        // Note: This workspace.textDocuments array only ever contains a single document or I would of used it instead.
+        
+        activeDocuments.forEach(doc =>{
```

Comment:
I suggest to include all opened documents, and all the lisp files included in current .prj file

Reply:
The only way I've found to get an accurate representation of what is truly open on-load is setting up the _workspace.onDidOpenTextDocument_ event and puppeteer the interface to cycle through all the available tabs. Which is **really** ugly, but our plugin isn't loaded until after the workspace has reloaded its previous state. 

This cycle-sync and I think a few other things about this update could probably use some new User-Options. I certainly wouldn't want this cycle to happen every time I hit F5, but I wouldn't mind it if it happened once at the start of my workday.
```typescript
const startingDocument = vscode.window.activeTextEditor.document.uri.fsPath;
	if (startingDocument.toUpperCase().slice(-4) === ".LSP"){
		do {
			await vscode.commands.executeCommand('workbench.action.nextEditor');
		} while (vscode.window.activeTextEditor.document.uri.fsPath !== startingDocument);
	}
```

Reply:
` our plugin isn't loaded until after the workspace has reloaded its previous state.`
I think our plugin is loaded when opening a LSP files, hit F5 key, see this [page ](https://code.visualstudio.com/api/references/activation-events)

Reply:
Thanks, I'll see if I can find something in that link to give me the document access, but right now our activate isn't called until after its too late. IE, If you close VSCode with a workspace active and 5 documents open, it will open back up with 5 documents tabs, but our extension has zero context about 4 of them until they are activated.

Reply:
I tried the all inclusive "*" to see if our Activate () would trigger soon enough to capture the workspace state reload and it didn't. Which kind of makes sense because [this page](https://vscode-docs.readthedocs.io/en/stable/extensionAPI/activation-events/) states that _Extensions are activated lazily in VS Code_.

I really don't like the notion of cycling the interface and that was actually the foundation of why I made that enhancement request for Import comment headers. If those existed, I really wouldn't care if the total scope of open documents were known/available because an explicit specification would ultimately be 10x better than including the defuns from some random document the user just happened to open, but may or may not be associated with something they care about in the active LSP.

Reply:
Can we reference other open source extension like python for vscode? Or search some help from the vscode slack https://vscode-dev-community.slack.com/

Reply:
I've tried finding a public invite/join link to that slack community, but I never did find one.

I do think our problem is somewhat unique due to the archaic nature of AutoLisp. The vast majority of languages wouldn't bothered trying to solve this issue because they don't need it. They already have some version of 'import' that clearly defines what they should be looking at. AutoLisp is a globalized free for all and we don't really have that luxury.

I've been looking at [this](https://github.com/DonJayamanne/pythonVSCode/) python repo off and on for a while. They have some interesting stuff worth paying attention to, but nothing I related to the problem of finding documents that were opened prior to extension activation. At least nothing stands out to me that I haven't already tried.

Reply:
I don't know why I didn't think of this sooner. I went and searched the vscode repo for slack and found the link there

Reply:
That slack channel will be a big help, but unfortunately all the chatter suggests this is long standing issue that hasn't been addressed in VSCode yet. They have both of these namespaces returning single objects regardless of what is open and it seems like I am not the only one who doesn't understand why these things are named something that they simply are not.
vscode.workspace.textdocuments
vscode.window.visualtexteditors

Reply:
Thanks for taking time to do researching! Maybe we leave some technical debt here, then do cleanup if in future find some methods to do it.

Reply:
Yep, that makes a lot more sense than implementing a deplorable workaround. I'll add some prominent TODO notes.

As soon as Sen-Real gets back to us on that one bug I'll probably be canceling this PR to go do a bunch of reconfiguration and scope it only to the context/document manager for now. Maybe this bug isn't a concern because it only fires on malformed (multi-line for sure) strings. Maybe we should just do a console log message, let it do what it can and fail where it does?